portability-worker/src/main/java/org/dataportabilityproject/worker/WorkerMain.java

```diff
@@ -18,17 +18,66 @@
 import com.google.common.util.concurrent.UncaughtExceptionHandlers;
 import com.google.inject.Guice;
 import com.google.inject.Injector;
+import org.dataportabilityproject.api.launcher.ExtensionContext;
+import org.dataportabilityproject.api.launcher.Logger;
+import org.dataportabilityproject.spi.transfer.extension.TransferExtension;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.ServiceLoader;
 
 /**
  * Main class to bootstrap a portabilty worker that will operate on a single job whose state
  * is held in WorkerJobMetadata.
  */
 public class WorkerMain {
-  public static void main(String[] args) throws Exception {
-    Thread.setDefaultUncaughtExceptionHandler(UncaughtExceptionHandlers.systemExit());
-    Injector injector = Guice.createInjector(new WorkerModule());
-    WorkerImpl worker = injector.getInstance(WorkerImpl.class);
-    worker.processJob();
-    System.exit(0);
-  }
+    public static void main(String[] args) throws Exception {
+        Thread.setDefaultUncaughtExceptionHandler(UncaughtExceptionHandlers.systemExit());
+
+        ExtensionContext extensionContext = getExtensionContext();
+
+        List<TransferExtension> extensions = new ArrayList<>();
+        ServiceLoader.load(TransferExtension.class).iterator().forEachRemaining(extensions::add);
```

Comment:
This is probably due to my lack of java knowledge. You mentioned that this ServiceLoader works pre-java9 by using META-INF/services files, are these automatically generated or something that we would need to check into git? 

I think @holachuy 's question above might also clear this one up: for V1, this would mean that we specify all the valid services in this file and just load them all in our jar. How would this look for a V2, where we selectively load the service classes?

Reply:
Yes, that's correct they work pre-Java 9. Here's an example:

http://www.logicbig.com/tutorials/core-java-tutorial/java-se-api/service-loader/

The idea is each jar/module extension would create a META-INF/services file. When the jar is on the classpath, ServiceLoader will find it's implementation. For uber/shaded jars, the Gradle plugin will concatenate the META-INF/services file so that multiple implementations are included in it.

Java 9 adds the ability to use module-info declarations instead of META-INF services, but we don't need to use this for now. For example:

module com.foo.myprovider {

    provides org.dataportabilityproject.transfer.spi.TransferExtension with com.foo.MyExtensionImpl;

} 

And a module-info that uses the extensions:

module com.foo.myworkerprocess {
    uses org.dataportabilityproject.transfer.spi.TransferExtension;
}

For simplicity, I think each extension can just keep META-INF/services files in its /resource directory. 

For now, the pre-Java9 approach should be OK but in the future, I think we are going to want a way to "assemble" images that's easier to use. For example, someone may want to create a worker process with there own set of extensions rolled up into a Docker image. There are number of different approaches we can probably take here once we have things working with the new APIs/SPIs.

Reply:
yeah, this will be interesting come java 9. I can imagine a situation where we have many permutations of different modular jars for the various import/export service combinations, and getting our worker pool (be it Kubernetes or something else) to spin up a worker with the correct image with the correct jar for each job could be complicated. But maybe that is what you have to do in order to get full "modularity" here where we are not pulling in code from unused libraries.

Reply:
For this case, I think we could keep things simple to start: all workers run all extensions and hence can support all combinations. 

In the future, if we want to support different scaling characteristics one way to do this would be to use Java 9 and create module layers that loaded modules dynamically based on environment variables. Kubernetes or some other orchestrator could launch an image and set the environment variables based on how it wants to balance processes. 

There's not much written on module layers (outside of the JPMS spec) yet but they are pretty powerful since they allow Java code to dynamically load (and unload) modules at runtime rather than having to specify the module path at startup. They can also be used to support versioning and isolation since a layer can hold one or more modules and be associated with a classloader.